drivers/clock_control/clock_control_rcar_cpg_mssr.c

```diff
@@ -59,71 +29,37 @@ static void cpg_write(const struct rcar_mssr_config *config,
 static void cpg_reset(const struct rcar_mssr_config *config,
 		      uint32_t reg, uint32_t bit)
 {
-	cpg_write(config, SRCR(reg), BIT(bit));
-	cpg_write(config, SRSTCLR(reg), BIT(bit));
-}
-
-static int cpg_core_clock_endisable(const struct device *dev,
-				    uint32_t module, uint32_t rate, bool enable)
-{
-	const struct rcar_mssr_config *config = dev->config;
-	uint32_t divider;
-	unsigned int key;
-	int ret;
+	uint16_t offset = soc_cpg_get_srcr_offset(reg);
 
-	/* Only support CANFD core clock at the moment */
-	if (module != CPG_CORE_CLK_CANFD) {
-		return -EINVAL;
-	}
-
-	key = irq_lock();
-
-	if (enable) {
-		if ((CANFDCKCR_PARENT_CLK_RATE % rate) != 0) {
-			__ASSERT(true, "Can not generate "
-				 "%u from CANFD parent clock", rate);
-			ret = -EINVAL;
-			goto unlock;
-		}
-		divider = (CANFDCKCR_PARENT_CLK_RATE / rate) - 1;
-		if (divider > CANFDCKCR_DIVIDER_MASK) {
-			__ASSERT(true, "Can not generate %u "
-				 "from CANFD parent clock", rate);
-			ret = -EINVAL;
-			goto unlock;
-		}
-		cpg_write(config, CANFDCKCR, divider);
-	} else {
-		cpg_write(config, CANFDCKCR, CANFDCKCR_CKSTP);
-	}
-
-unlock:
-	irq_unlock(key);
-	return 0;
+	cpg_write(config, offset, BIT(bit));
+	cpg_write(config, SRSTCLR(reg), BIT(bit));
 }
 
-static int cpg_rmstp_clock_endisable(const struct device *dev,
-				     uint32_t module, bool enable)
+static int cpg_mstp_clock_endisable(const struct device *dev,
+				    uint32_t module, bool enable)
 {
 	const struct rcar_mssr_config *config = dev->config;
 	uint32_t reg = module / 100;
 	uint32_t bit = module % 100;
 	uint32_t bitmask = BIT(bit);
 	uint32_t reg_val;
-
-	__ASSERT((bit < 32) && reg < ARRAY_SIZE(rmstpsr),
-		 "Invalid module number for cpg clock: %d", module);
+	uint16_t offset;
 
 	unsigned int key = irq_lock();
 
-	reg_val = sys_read32(config->base_address + RMSTPSR(reg));
+	if (soc_cpg_get_mstpcr_offset(module, &offset)) {
+		irq_unlock(key);
+		return 1;
+	}
+
```

Comment:
you can lock the irq after getting the register offset. please also return the value of soc_cpg_get_mstpcr_offset in case of error.

int ret;
unsigned int key;

ret = soc_cpg_get_mstpcr_offset(module, &offset);
if (ret < 0) {
return ret;
}
key = irq_lock();